spec-tests/src/test/java/org/candlepin/spec/products/ProductVersioningSpecTest.java

```diff
@@ -0,0 +1,664 @@
+        assertThat(prods).hasSize(1);
+        ProductDTO prod6 = prods.get(0);
+        assertNotEquals(prod6.getUuid(), prod1.getUuid());
+        assertNotEquals(prod6.getUuid(), prod2.getUuid());
+        assertEquals(prod6.getUuid(), prod3.getUuid());
+        assertNotEquals(prod6.getUuid(), prod4.getUuid());
+        assertEquals(prod6.getUuid(), prod5.getUuid());
+    }
+
+    @Test
+    public void shouldCleanupOrphansWithoutInterferingWithNormalActions() throws Exception {
+        // NOTE:
+        // This test takes advantage of the immutable nature of products with the in-place update branch
+        // disabled. If in-place updates are ever reenabled, we'll need a way to generate large numbers
+        // of orphaned products for this test.
+
+        OwnerDTO owner1 = ownerApi.createOwner(Owners.random());
+        OwnerDTO owner2 = ownerApi.createOwner(Owners.random());
+
+        String prefix = "test-product-";
+        int offset = 10000;
+        int length = 100;
+
+        // Repeat this test a few(ish) times to hopefully catch any synchronization error
+        for (int i = 0; i < 10; i++) {
+            final int base = offset;
+            List<String> o1Uuids = new ArrayList<>();
+            List<String> o2Uuids = new ArrayList<>();
+
+            // Create a bunch of dummy products
+            for (int j = base; j < base + length - 1; j++) {
+                String id = String.format("%s%d", prefix, j);
+                // create product and immediately update it to generate an orphaned product
+                ownerProductApi.createProductByOwner(owner1.getKey(), new ProductDTO().name(id).id(id));
+            }
+
+            // Attempt to update and create new products to get into some funky race conditions with
+            // convergence and orphanization
+            Thread updater = new Thread(() -> {
+                for (int j = base; j < base + length - 1; j++) {
+                    String id = String.format("%s%d", prefix, j);
+                    // create product and immediately update it to generate an orphaned product
+                    ProductDTO prod = ownerProductApi.updateProductByOwner(owner1.getKey(), id,
+                        new ProductDTO().name(id + "-update"));
+                    o1Uuids.add(prod.getUuid());
+                }
+            });
+            updater.start();
+            Thread generator = new Thread(() -> {
+                for (int j = base; j < base + length - 1; j++) {
+                    String id = String.format("%s%d", prefix, j);
+                    ProductDTO prod = ownerProductApi.createProductByOwner(owner2.getKey(),
+                        new ProductDTO().name(id).id(id));
+                    o2Uuids.add(prod.getUuid());
+                }
+            });
+            generator.start();
+
+            sleep(1000);
+            AsyncJobStatusDTO job = client.jobs().scheduleJob("OrphanCleanupJob");
+            updater.join();
```

Comment:
We should add an upper-bound on these joins so we don't wait indefinitely in the event something gets stuck somewhere. In the content version of these tests we went with 90s.